src/engine/GroupBy.cpp

```diff
@@ -740,81 +771,70 @@ bool GroupBy::computeGroupByForSortedSubtree(IdTable* result,
 
   for (size_t i = 0; i < numBlocks; ++i) {
     evaluationContext._beginIndex = i * blockSize;
     evaluationContext._endIndex = std::min((i + 1) * blockSize, subresult.size());
 
     auto currentBlockSize = evaluationContext._endIndex - evaluationContext._beginIndex;
 
-    std::vector<sparqlExpression::detail::IntOrDouble> resultValues;
+    vector<sparqlExpression::SparqlExpression*> results;
 
-    // Is this slower than emplace_back?
-    // Calling IntOrDouble constructor twice VS. resizing of vector on the go
-    resultValues.resize(currentBlockSize);
+    size_t aggregateIndex = 0;
+    auto currentRowIndex = rowIndex;
 
-    // Evaluate child expression on block
-    auto expr = aggregates.at(0)._expression.getPimpl();
-    auto exprChildren = expr->children();
+    for (auto & aggregate : aggregates) {
+      // Evaluate child expression on block
+      auto expr = aggregate._expression.getPimpl();
 
-    sparqlExpression::ExpressionResult expressionResult =
-        exprChildren[0]->evaluate(&evaluationContext);
+      // Only allow AVG for now
+      if (dynamic_cast<sparqlExpression::AvgExpression*>(expr) == nullptr)
+        AD_FAIL();
 
-    auto visitor = [&]<sparqlExpression::SingleExpressionResult T>(
-                       T&& singleResult) mutable {
-      auto generator = sparqlExpression::detail::makeGenerator(
-          std::forward<T>(singleResult), currentBlockSize,
-          &evaluationContext);
+      auto exprChildren = expr->children();
 
-      using NVG = sparqlExpression::detail::NumericValueGetter;
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
 
-      size_t k = 0;
-      for (auto val : generator) {
-        sparqlExpression::detail::NumericValue numVal = NVG()(val, &evaluationContext);
+      auto visitor = [&]<sparqlExpression::SingleExpressionResult T>(
+                         T&& singleResult) mutable {
+        auto generator = sparqlExpression::detail::makeGenerator(
+            std::forward<T>(singleResult), currentBlockSize,
+            &evaluationContext);
 
-        if (const int64_t* pInt = std::get_if<int64_t>(&numVal))
-          resultValues[k] = *pInt;
-        else if (const double* pDouble = std::get_if<double>(&numVal))
-          resultValues[k] = *pDouble;
-        else
-          AD_FAIL();
-        ++k;
-      }
-      overallResults += k;
-    };
+        using NVG = sparqlExpression::detail::NumericValueGetter;
 
-    std::visit(visitor, std::move(expressionResult));
+        for (const auto & val : generator) {
+          sparqlExpression::detail::NumericValue numVal = NVG()(val, &evaluationContext);
+          auto id = subresult(rowIndex, columnIndex);
 
-    for (size_t j = 0; j < currentBlockSize; ++j) {
-      auto rowIndex = i * blockSize + j;
-      auto id = subresult(rowIndex, columnIndex);
-      map[id].increment(resultValues[j]);
-    }
-  }
+          if (map[id].size() <= aggregateIndex)
```

Comment:
We probably can do the lookup in the maps only once for all the aggregates, but that can also be in a follow-up PR.